Implement port forwarding.

iciclespider opened the pull request:
Looking for feed back if this looks on track as a way of implementing port forwarding.

See the example and unittests added in this PR https://github.com/kubernetes-client/python/pull/1237 for usage.

https://github.com/kubernetes-client/python/issues/166

iciclespider commented:
/check-cla

iciclespider commented:
/check-cla

iciclespider commented:
/assign @mbohlool

yliaog commented:
/assign

yliaog commented:
thanks for the PR. it looks good overall. 

could you please split this PR into two PRs, the first PR is for refactoring the existing code to make it reusable for port forwarding, the second PR is for port forwarding.

this makes it easier to review, and also easier to rollback or fix the portforward PR in case it has any issue.

iciclespider commented:
@yliaog  PR to refactor existing implementation in preparation for implementing port forwarding: #211

iciclespider commented:
@yliaog I rebased this PR against the merged master with the refactored code. I also updated the PR in https://github.com/kubernetes-client/python/pull/1237 with a `pod_portforward.py` example.

palnabarun commented on stream/ws_client.py:
```diff
@@ -302,3 +441,34 @@ def websocket_call(configuration, _method, url, **kwargs):
+
+    query_params = kwargs.get("query_params")
+
+    ports = []
+    for key, value in query_params:
+        if key == 'ports':
+            for port in value.split(','):
+                try:
+                    port = int(port)
+                    if not (0 < port < 65536):
+                        raise ValueError
+                    ports.append(port)
+                except ValueError:
+                    raise ApiValueError("Invalid port number `" + str(port) + "`")
+    if not ports:
+        raise ApiValueError("Missing required parameter `ports`")
```
I think this logic can be abstracted out to another function with a simpler implementation like below.

```
def _valid_ports(ports):
    _ports = []

    for p in ports:
        try:
            p = int(p)
        except ValueError:
            raise ApiValueError("Invalid port number `{}`. Port number must be an integer".format(p))

        if (p < 0) or (p > 65536):
            raise ApiValueError("Invalid port number `{}`. Port number must be between 0 and 65536".format(p))

        _ports.append(p)

    return _ports

ports = query_params.get("ports", "")
if not ports:
    raise ApiValueError("Missing required parameter `ports`")

ports = ports.split(",")
ports = _valid_ports(ports)

# There is no check required here, since any unwanted port value would be checked in _valid_ports itself.
# We can be assured that we get a non-empty port list here.
```

iciclespider replied:
This line will not work:
```
ports = query_params.get("ports", "")
```
query_params is not a dictionary, it an a list of two element tuples.

palnabarun commented:
/assign

yliaog commented on stream/ws_client.py:
```diff
@@ -302,3 +441,34 @@ def websocket_call(configuration, _method, url, **kwargs):
+    ports = []
+    for key, value in query_params:
+        if key == 'ports':
+            for port in value.split(','):
```
ports should be list of local_port:remote_port pairs

yliaog replied:
https://github.com/kubernetes/client-go/blob/master/tools/portforward/portforward.go

is the implementation for client-go

iciclespider replied:
Good catch. I'll adjust this parsing logic.

yliaog commented on stream/ws_client.py:
```diff
@@ -225,6 +227,143 @@ def close(self, **kwargs):
+    def _proxy(self):
+        channel_ports = []
+        channel_initialized = []
+        python_ports = {}
```
what about call it local_port?

yliaog commented on stream/ws_client.py:
```diff
@@ -225,6 +227,143 @@ def close(self, **kwargs):
+        def __init__(self, ix, number):
+            self.number = number
+            self.channel = bytes([ix * 2])
+            s, self.python = socket.socketpair(socket.AF_UNIX, socket.SOCK_STREAM)
```
what about renaming self.python to self.local?

yliaog commented on stream/ws_client.py:
```diff
@@ -225,6 +227,143 @@ def close(self, **kwargs):
+            self.number = number
+            self.channel = bytes([ix * 2])
+            s, self.python = socket.socketpair(socket.AF_UNIX, socket.SOCK_STREAM)
+            self.socket = self._Socket(s)
```
what about renaming self.socket to self.remote_xxx?

yliaog replied:
actually, why is sockepair needed? only one socket for each local port is needed, no?

iciclespider replied:
There are the following `socket`s involved:

1. The websocket's socket used to receive and send data to kubernetes. This is the `PortForward.websocket.sock` field. There is only one such socket involved.
2. The socket given to the running python code that the python code uses to send and receive data. This is currently what the `_Port.socket` field is. I called it `socket` because it is the socket supplied to the python code.
3. The socket used by the `PortForward._proxy` method to receive and send data to the python code that is using the `_Port.socket` socket. This is currently what the `_Port.python` code is. I called it `python` because it is the socket that `_proxy` is using to interact with the python code with.

The flow of data being sent from the python to kubernetes using this is:

* The python send data using the `_Port.socket` it has been supplied with.
* The `PortForward._proxy` method receives the above data using the `_Port.python` socket.
* The `PortForward._proxy` method sends the received data to `PortForward.websocket.sock` using the websocket channel that that port is associated with and using the websocket library to format the data.

The flow of data being sent from kubernetes to python using this is:

* The `PortForward._proxy` method receives data from kubernetes on the `PortForward.websocket.sock` socket and using the websocket library
* The `PortForward._proxy` method` sends the received data to the `_Port.python` socket associated with the websocket channel the data arrived on.
* The python code receives the above data on it's `_Port.socket` socket.

Hopefully the above makes clear what the socketpair is accomplishing. It is used to enable translating between the "raw" socket protocol the python is using, and the websocket based protocol that kubernetes is using, all without the python knowing what is going on.

So this still leaves what to call each end of the socketpair. I chose `_Port.socket` for the name of the socket used by the running python application to use as it's "socket". And I chose `_Port.python` for the name of the socket used by `PortForward._proxy` because it is using that socket to interact with the running python application.

I am will call them any name you prefer.

yliaog replied:
thanks for the clarification. here is my understanding, take a simple example, portforward from localhost:8080 to remote pod nginx port 80.

* there is just one websocket PortForward.websocket.sock used to send to/recv from pod nginx:80

* there is one local socket that binds to localhost:8080

1. listen on the socket (localhost:8080), wait for connection request
2. when there is a connection request, accept it, open a new connection, use that connection to  communicate with remote pod nginx:80 via the websocket

there could be multiple connection requests, hence multiple connections could be opened to handle them.

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/portforward/portforward.go is the client-go's implementation.

i don't quite see how socketpair fits into the above flow.

iciclespider replied:
Please look at the code example in this PR https://github.com/kubernetes-client/python/pull/1237

This implementation does not expose a port for other applications to connect to. Instead, it allows a python application to interact with the pod ports. There is not attempt to expose a port that other non-python executables can access in the same way kubectl exposes a port for other applications to use.

iciclespider replied:
I 100% agree. I got confused by your confusion! After I implemented the local port/remote port handling, I realized that you did not understand that an actual local port was not used, and it finally dawned on me that this is really not necessary.

I'll back out that local/remote port work and address your other comments.

iciclespider commented on stream/stream.py:
```diff
@@ -17,9 +17,12 @@
     method and call the supplied Kubernetes API method with that in place."""
+    if force_kwargs:
+        for kwarg, value in force_kwargs.items():
+            kwargs[kwarg] = value
```
It must be supplied to the `connect_get_namespaced_pod_portforward` when it is called in the method so that the kubernetes client implementation between calling `connect_get_namespaced_pod_portforward` and calling `ws_client.portforward_call` returns the raw response, and does not try to return the json'ed version of what it expects to be a typical json reply.

iciclespider commented on stream/ws_client.py:
```diff
@@ -225,6 +227,143 @@ def close(self, **kwargs):
+            def __getattr__(self, name):
+                return getattr(self._socket, name)
+
+            def setsockopt(self, level, optname, value):
```
`http.client` makes this call right here: https://github.com/python/cpython/blob/8ca8a2e8fb068863c1138f07e3098478ef8be12e/Lib/http/client.py#L938

The code in this PR encounters this situation when socket.client_connection is monkey patched to return this connected socket in response to the `<pod-name>.<namespace>.kubernetes` dns names. This then allows the python code to use `http.client` or use `urllib` to make http requests that get routed to kubernetes pods.

It turns out that a socket returned from socketpair does not support the option being checked for below and throws an exception:
```
                if level == socket.IPPROTO_TCP and optname == socket.TCP_NODELAY:
```
Hence, the entire purpose of this `class _Socket` is to address this very issue. Otherwise the socket returned from socketpair could have been directly returned.

yliaog replied:
ok

iciclespider commented on stream/ws_client.py:
```diff
@@ -225,6 +227,143 @@ def close(self, **kwargs):
+            channel_ports.append(port)
+            channel_initialized.append(False)
+            channel_ports.append(port)
+            channel_initialized.append(False)
```
This is intentional to setup the data channel and the error channel. I will add a comment to make this clear.

iciclespider commented on stream/ws_client.py:
```diff
@@ -225,6 +227,143 @@ def close(self, **kwargs):
+            raise ValueError("Invalid port number")
+        return self.ports[port_number].socket
+
+    def error(self, port_number):
```
I just pushed the change to `error_channel`.

However, a little clarification. This is not really a "channel" that can send and receive data. From the client perspective, it is read only and cannot be written to. And the only time it is written to by kubernetes is when there is an error/exception case on the server.

See: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cri/streaming/portforward/websocket.go#L184-L197

yliaog commented on stream/ws_client.py:
```diff
@@ -302,3 +441,56 @@ def websocket_call(configuration, _method, url, **kwargs):
+
+    query_params = kwargs.get("query_params")
+
+    ports = []
```
so ports is a list, if the query_params is ports="80,80,80", ports would have three dups, better to remove dups up front

iciclespider commented on stream/ws_client.py:
```diff
@@ -225,6 +227,143 @@ def close(self, **kwargs):
+            self.local_ports[local_remote[0]] = self._Port(ix, local_remote[1])
+        threading.Thread(
+            name="Kubernetes port forward proxy", target=self._proxy, daemon=True
+        ).start()
```
One such thread for each PortForward instance. It terminates after either side is closed once all pending data is flushed. I will comment it.

iciclespider commented on stream/ws_client.py:
```diff
@@ -225,6 +227,143 @@ def close(self, **kwargs):
+                    else:
+                        port.python.close()
+                        rlist.remove(s)
+                        if len(rlist) == 1:
```
I reworked the detection of how and when it is time to terminate this _proxy thread.

iciclespider commented:
@yliaog I submitted another commit for review.

yliaog commented on stream/ws_client.py:
```diff
@@ -285,42 +303,44 @@ def setsockopt(self, level, optname, value):
+                        all_closed = False
+                    else:
+                        if self.websocket.connected:
+                            rlist.append(port.python)
```
so currently rlist.append(port.python) is called only when both (port.data is None) and (self.websocket.connected is True)

IMO, whether port.data is None or not should not be used to decide whether to read from port.python, what do you think?

iciclespider replied:
Good catch, I noticed that myself and was already reworking this code.

yliaog commented on stream/ws_client.py:
```diff
@@ -285,42 +303,44 @@ def setsockopt(self, level, optname, value):
+                            all_closed = False
+                        else:
+                            port.python.close()
+            if all_closed and (not self.websocket.connected or not kubernetes_data):
```
if all_closed, and kubernetes_data, where is the 'kuberentes_data' going to be sent to?

iciclespider replied:
`all_closed` is just all the local ports. I renamed this variable to be `local_all_closed`.

yliaog replied:
well, for the following condition:

 * all_closed, and self.websocket.connected,  and kubernetes_data *

the 'if' condition is line 337 would be false, hence, websokcet won't close, and the fuction won't return.

but since all_closed is true, even though kubernetes_data is not None, there is no socket to receive that kubernetes_data, there seems to be no point to keep the thread running.

yliaog replied:
i think the scenario could happen when:

immediately after python application send some request, it closes the socket, instead of waiting to receive any response.

iciclespider replied:
> but since all_closed is true, even though kubernetes_data is not None, there is no socket to receive that kubernetes_data, there seems to be no point to keep the thread running.

There is a socket to receive the data, the PortFoward.websocket.sock socket. That socket has not yet been closed by virtue of the fact that PortForward.websocket.connected is still True.

iciclespider replied:
I reworked that if to instead be:
```
            if local_all_closed and not (self.websocket.connected and kubernetes_data):
```
I think it is clearer that way.

yliaog replied:
ok.

iciclespider reviewed (commented):
@yliaog I pushed a commit that addresses the comments.

yliaog commented on stream/ws_client.py:
```diff
@@ -225,6 +227,174 @@ def close(self, **kwargs):
+                            ABNF.OPCODE_BINARY,
+                        ).format()
+                    else:
+                        port.python.close()
```
previously, there was the condition before the line above

                 if not port.data:

why is it removed?

iciclespider replied:
I removed it because the socket has been closed by the python application. This also makes it symmetrical with the closing down of the websocket half. If there is `kubernetes_data` left over when the websocket becomes `not connected`, the `kubernetes_data` is not sent. Same with the python socket. There is no longer a functional socket to send the remaining data on.

yliaog replied:
sounds good.

iciclespider commented on stream/ws_client.py:
```diff
@@ -225,6 +227,174 @@ def close(self, **kwargs):
+                            wlist.append(port.python)
+                            local_all_closed = False
+                        else:
+                            port.python.close()
```
> when there is port.error, as soon as port.data is flushed, port.python is closed (even when websocket.connected is true). i think that is correct, but want to double confirm.

That is intentional. The only time when a message arrives on the error channel is when the data channel either did not initialize or is not longer working.

> also port.error is not delivered to the python app, only port.data is delivered. it's worth to mention this behavior in the documentation.

There is the `PortForward.error(<port-number)` call to check if an error occurred.

yliaog replied:
yes, better to have it in the documentation. but i guess the examples / e2e tests would make it clear too.

yliaog commented:
thanks for the PR, really appreciate it.

/lgtm
/approve
